simpledotorg/simple-server
# Allow mapping business identifiers with facilities from the facility bulk upload page

**Story card:** [sc-10741](https://app.shortcut.com/simpledotorg/story/10741/implement-facility-mapping-scripts)

## Because

We will have new stakeholder organisations in Bangladesh sending us their data in order to use the Simple dashboard for reporting. This change is an important first step that will let us accept external organisations that want to export their data and see the reports on the Simple dashboard. What we are doing is adding a new column in the facility bulk upload template that will let organisations map their business identifiers to the facilities they are uploading.

## This addresses
* Changes to the bulk upload template to have a business identifiers column.
* Creation of a facility business identifier when a row has the aforementioned column populated.
* Validations of the uploaded data, to ensure no duplicates.

Not covered in this story:
* Editing or deleting the business identifiers.

## Test instructions

* Get simple-server with sidekiq up and running.
* Create a new organisation.
* Go to the bulk add facilities page: http://localhost:3000/admin/facilities/upload (or open it from the facility settings page)
* Follow the instructions in that page, ensure that the districts and blocks for the new org is set up.
* Download the CSV template, add some dummy data. Leave some of the business identifiers blank.
* Ensure that the facilities were added correctly.
* In Rails console, run `FacilityBusinessIdentifier.where(identifier_type: "external_org_facility_id")`
    * Ensure that there is an identifier for every facility where the business identifier column was populated.
